### PR TITLE
Bob

### PR DESCRIPTION
**Pull request by amymhaddad**

### CODE REVIEW

File: python/bob/bob.py [hunk 1]
```diff
@@ -8,36 +11,25 @@
...
-    if shouting:
-        return RESPONSES['shout']
-    if asking_question:
+    if question(phrase):
         return RESPONSES['question']
-    return RESPONSES['default']
-
-def is_silent(phrase):
-    """Is the phrase blank?"""
+    if all_caps(phrase):
```
Comment (paulghaddad) [inline python/bob/bob.py]: Consider using the `if-elif-else` construct instead of a bunch of `if`s.

File: python/bob/bob.py [hunk 2]
```diff
@@ -8,36 +11,25 @@
...
+        return RESPONSES['shout']
+    if statement(phrase):
+        return RESPONSES['default']
 
-    return phrase == ""
+def silent(phrase):
+    return re.match(r'^\s*$', phrase)
```
Comment (paulghaddad) [inline python/bob/bob.py]: Wow, impressed! One thing you can do here is to extract the pattern into a well named constant. Something like:

```
SILENCE_REGEX = r'^\s*$'
```

That way it's clear what this regex does and you can reuse it (and maybe compile it) later.

File: python/bob/bob.py [hunk 3]
```diff
@@ -8,36 +11,25 @@
...
 
-def is_shouting(phrase):
-    """Is the phrase in all capital letters?"""
+def statement(phrase):
+    return re.search(r'[\w*\W*\s*]', phrase)
```
Comment (paulghaddad) [inline python/bob/bob.py]: Consider extracting this regex out, similar to my comment above.

File: python/bob/bob.py [hunk 4]
```diff
@@ -8,36 +11,25 @@
...
 
+def all_caps(phrase):
     return phrase.isupper()
 
-def is_asking_question(phrase):
-    """Is the phrase a question?"""
-
-    return phrase.endswith("?")
+def question(phrase):
+    return re.search(r'^[\w|\d|\W]*\?\s*$', phrase)
```
Comment (paulghaddad) [inline python/bob/bob.py]: In this case, I think the original code is simpler to understand.

**Comment by amymhaddad** [comment]
@paulghaddad Bob is updated! I hope the right version is showing for you. The version you made edits to was an earlier version. Plus I added a tag to the commit, so hopefully the 


```

"""Test to see Bob's responses"""

import re

RESPONSES = {
    'question': 'Sure.',
    'shout': 'Whoa, chill out!',
    'shout_question': "Calm down, I know what I'm doing!",
    'silence': 'Fine. Be that way!',
    'default': 'Whatever.',
    }

SILENCE = r'^\s*$'
ALL_CAPS = r'[A-Z]+[\d!]*?\Z'
SHOUT_QUESTION = r'([A-Z\s*]+)([A-Z]+\?\Z)'
QUESTION = r'^[\w|\d|\W]*\?\s*$'

def hey(phrase):
    """What is Bob's response?"""

    if silent(phrase):
        return RESPONSES['silence']
    elif shout_question(phrase):
        return RESPONSES['shout_question']
    elif question(phrase):
        return RESPONSES['question']
    elif all_caps(phrase):
        return RESPONSES['shout']
    return RESPONSES['default']

def silent(phrase):
    return re.match(SILENCE, phrase)

def all_caps(phrase):
    return re.search(ALL_CAPS, phrase)

def shout_question(phrase):
    return re.match(SHOUT_QUESTION, phrase)
  
def question(phrase):
    return re.search(QUESTION, phrase)

```

**Pull request closed without merging**